compiler/lib/flow.ml

```diff
@@ -360,53 +360,25 @@ let the_def_of info x =
         x
   | Pc c -> Some (Constant c)
 
-(* If [constant_identical a b = true], then the two values cannot be
-   distinguished, i.e., they are not different objects (and [caml_js_equals a b
-   = true]) and if both are floats, they are bitwise equal. *)
-let constant_identical ~(target : [ `JavaScript | `Wasm ]) a b =
-  match a, b, target with
-  | Int i, Int j, _ -> Targetint.equal i j
-  | Float a, Float b, `JavaScript -> Int64.equal a b
-  | Float _, Float _, `Wasm -> false
-  | NativeString a, NativeString b, `JavaScript -> Native_string.equal a b
-  | NativeString _, NativeString _, `Wasm ->
-      false
-      (* Native strings are boxed (JavaScript objects) in Wasm and are
-          possibly different objects *)
-  | String a, String b, `JavaScript -> Config.Flag.use_js_string () && String.equal a b
-  | String _, String _, `Wasm ->
-      false (* Strings are boxed in Wasm and are possibly different objects *)
-  | Int32 _, Int32 _, `Wasm ->
-      false (* [Int32]s are boxed in Wasm and are possibly different objects *)
-  | Int32 a, Int32 b, `JavaScript -> Int32.equal a b
-  | NativeInt _, NativeInt _, `Wasm ->
-      false (* [NativeInt]s are boxed in Wasm and are possibly different objects *)
-  | NativeInt a, NativeInt b, `JavaScript -> Int32.equal a b
-  (* All other values may be distinct objects and thus different by [caml_js_equals]. *)
-  | Int64 _, Int64 _, _ -> false
-  | Tuple _, Tuple _, _ -> false
-  | Float_array _, Float_array _, _ -> false
-  | (Int _ | Float _ | Int64 _ | Int32 _ | NativeInt _), _, _ -> false
-  | (String _ | NativeString _), _, _ -> false
-  | (Float_array _ | Tuple _), _, _ -> false
-
-let the_const_of ~target info x =
+let the_const_of ~eq info x =
   match x with
   | Pv x ->
       get_approx
         info
         (fun x ->
-          match info.info_defs.(Var.idx x), target with
-          | Expr (Constant ((Float _ | Int _ | NativeString _) as c)), _ -> Some c
-          | Expr (Constant ((Int32 _ | NativeInt _) as c)), `JavaScript -> Some c
-          | Expr (Constant (String _ as c)), _ when Config.Flag.safe_string () -> Some c
-          | Expr (Constant c), _ ->
-              if Var.ISet.mem info.info_possibly_mutable x then None else Some c
+          match info.info_defs.(Var.idx x) with
+          | Expr
+              (Constant
+                 ((Float _ | Int _ | Int32 _ | Int64 _ | NativeInt _ | NativeString _) as
+                  c)) -> Some c
+          | Expr (Constant c)
+            when Config.Flag.safe_string ()
+                 || not (Var.ISet.mem info.info_possibly_mutable x) -> Some c
```

Comment:
```suggestion
          | Expr (Constant (String _ as c)), _ when Config.Flag.safe_string () -> Some c
          | Expr (Constant c) -> 
             if  Var.ISet.mem info.info_possibly_mutable x then None else Some c
```

Reply:
I think neither code is correct.
When `Config.Flag.safe_string ()` is true, all constants are really constant.
If `Var.ISet.mem info.info_possibly_mutable x` if false, the root of the constant is not possibly modified. So, if it is a string, it is a constant string. If it's a tuple, we are not sure.